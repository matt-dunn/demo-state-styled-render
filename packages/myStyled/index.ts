/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

export { myStyled as default } from "./myStyled";

export { css, cssWithProps } from "./css";
