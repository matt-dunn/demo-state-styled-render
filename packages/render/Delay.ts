/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {AnyNode, Node} from "./";
import { useEffect, useState } from "./hooks";

type DelayProps = {
  delay?: number;
  children: Node;
};

export const Delay = ({ delay = 500, children }: DelayProps): AnyNode => {
  const [waiting, setWaiting] = useState(delay !== 0);

  useEffect(() => {
    if (waiting) {
      const t = setTimeout(() => {
        setWaiting(false);
      }, delay);

      return () => {
        clearTimeout(t);
      };
    }
  }, [delay]);

  return waiting ? null : children;
};
