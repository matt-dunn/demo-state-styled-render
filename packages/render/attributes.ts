/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { AnyFunc, HTMLElementMap, Props } from "./types";
import { propertyMap } from "./utils";

export const setAttribute = (
  element: HTMLElementMap,
  name: string,
  value: string | AnyFunc
) => {
  const propertyName = propertyMap(name);

  if (typeof value === "function") {
    element[propertyName] = value;
  } else {
    if (value) {
      element.setAttribute(propertyName, value);
    } else {
      element.removeAttribute(propertyName);
    }
    element[propertyName] = value;
  }
};

export const setAttributes = (element: HTMLElement, props: Props) =>
  Object.keys(props || {}).forEach((name) =>
    setAttribute(element, name, props[name])
  );

export const removeAttribute = (element: HTMLElement, name: string) =>
  element.removeAttribute(propertyMap(name));

export const updateAttribute = (
  element: HTMLElement,
  name: string,
  value: string,
  prevValue: string
) => {
  if (value !== prevValue) {
    if (value === undefined) {
      removeAttribute(element, name);
    } else if (!prevValue || value !== prevValue) {
      setAttribute(element, name, value);
    }
  }
};

export const updateAttributes = (
  element: HTMLElement,
  props: Props,
  prevProps: Props = {}
) =>
  Object.keys({ ...props, ...prevProps }).forEach((name) =>
    updateAttribute(element, name, props[name], prevProps[name])
  );
