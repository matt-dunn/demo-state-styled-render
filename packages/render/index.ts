/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

export { Mx as default } from "./render";

export { Node, Children, FC, HTMLElementMap } from "./types";

export { classnames } from "./utils";

export { jsx, jsxFrag, createElement } from "./element";

export * from "./hooks";
