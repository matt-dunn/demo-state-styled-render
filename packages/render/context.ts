/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { Node } from "packages/render";

import { looseRef } from "./utils";

export const getValueSymbol = Symbol("getValueSymbol");

export type Context<T = any> = {
  [getValueSymbol]: () => T | undefined;
  Provider: ({ value, children }: { value: T; children: Node }) => Node;
};

export const createContext = <T = any>(defaultValue?: T): Context<T> => {
  const currentValue = looseRef(defaultValue);

  return {
    [getValueSymbol]: () => currentValue.current,
    Provider: ({ value, children }) => {
      if (currentValue.current !== value) {
        currentValue.current = value;
      }

      return children;
    },
  };
};
