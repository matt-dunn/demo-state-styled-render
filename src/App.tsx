/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

/** @jsx jsx **/

import { jsx } from "packages/render";
import { StoreProvider } from "packages/state";
import { lazy } from "packages/render/lazy";

import { getStore } from "./store";

import { Header } from "./components/Header";

const store = getStore({
  todos: [
    { id: "1", text: "Do something", complete: true },
    { id: "2", text: "Do something else", complete: false },
  ],
});

const Todos = lazy(
  () => import("./containers/Todos"),
  (module) => module.Todos
);

export const App = () => (
  <StoreProvider store={store}>
    <div className="container-lg container-main">
      <Header />
      <article>
        <header>
          <p>Simple vanilla, dependency free implementation of:</p>
          <ul>
            <li>
              <a
                href="https://github.com/matt-dunn/demo-state-styled-render/tree/master/packages/render"
                target="_blank"
                rel="noopener noreferrer"
              >
                JSX functional component rendering with simple hooks
              </a>
            </li>
            <li>
              <a
                href="https://github.com/matt-dunn/demo-state-styled-render/tree/master/packages/state"
                target="_blank"
                rel="noopener noreferrer"
              >
                Redux style state management
              </a>
            </li>
            <li>
              <a
                href="https://github.com/matt-dunn/demo-state-styled-render/tree/master/packages/myStyled"
                target="_blank"
                rel="noopener noreferrer"
              >
                A Styled Component implementation
              </a>
            </li>
          </ul>
        </header>

        <p>
          Intended as an illustration of how the basics and internals of React /
          Redux / Styled Components work under the hood and{" "}
          <strong>not intended for production use</strong>.
        </p>
      </article>
      <Todos />
    </div>
  </StoreProvider>
);
