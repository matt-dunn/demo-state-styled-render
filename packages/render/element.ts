/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {
  AnyNode,
  Children,
  Key,
  Node,
  NODE_TYPE_FRAGMENT,
  NodeType,
  Props,
} from "./types";
import { isNode } from "./utils";
import { setAttributes, updateAttributes } from "./attributes";

export const jsx = (
  type: NodeType,
  props: Props = {},
  ...children: Children
): Node => {
  const { key, ...rest } = props || {};

  return typeof type === "function"
    ? type({
        ...rest,
        children: children?.length === 1 ? children[0] : children,
        key,
      })
    : { type, props: rest, children, key };
};

export const jsxFrag = ({
  children,
  ...props
}: {
  children: Children;
}): Node & { key: Key } => ({
  type: NODE_TYPE_FRAGMENT,
  props: props,
  children: children,
  key: null,
});

(global as any).jsx = jsx;
(global as any).jsxFrag = jsxFrag;

export const createElement = (
  type: NodeType,
  props: Props = {},
  children?: Children | Node
): Node => {
  if (Array.isArray(children)) {
    return jsx(type, props, ...children);
  } else if (children) {
    return jsx(type, props, children);
  } else {
    return jsx(type, props);
  }
};

const createDocumentElement = (node: AnyNode): HTMLElement | Text => {
  if (isNode(node) && typeof node.type === "string") {
    const element = document.createElement(node.type);

    setAttributes(element, node?.props);

    flattenChildren(node.children)
      .map(createDocumentElement)
      .forEach(element.appendChild.bind(element));

    return element;
  } else if (typeof node === "string" || typeof node === "number") {
    return document.createTextNode(node.toString());
  }

  throw new TypeError(
    `Unable to render invalid node ${JSON.stringify(node).substr(0, 20)}`
  );
};

const hasChanged = (node: AnyNode, prevNode: AnyNode) =>
  typeof node !== typeof prevNode ||
  (typeof node === "string" && node !== prevNode) ||
  (isNode(node) && node.type) !== (isNode(prevNode) && prevNode.type);

const updateChildren = (
  element: HTMLElement,
  children: Children,
  prevChildren: Children,
  index: number
) => {
  const nodeChildrenLength = children.length;
  const prevNodeChildrenLength = prevChildren.length;
  for (
    let childIndex = 0;
    childIndex < nodeChildrenLength || childIndex < prevNodeChildrenLength;
    childIndex++
  ) {
    updateTree(
      element.childNodes[index] as HTMLElement,
      children[childIndex],
      prevChildren[childIndex],
      childIndex
    );
  }
};

const flattenChildren = (children: Children = []): Children =>
  children.reduce((children, child) => {
    if (child.type === NODE_TYPE_FRAGMENT) {
      return [...children, ...flattenChildren(child.children)];
    }
    return [...children, child];
  }, [] as Children);

export const updateTree = (
  element: HTMLElement,
  node: Node,
  prevNode?: Node,
  index = 0
) => {
  if (prevNode === undefined) {
    element.appendChild(createDocumentElement(node));
  } else if (node === undefined) {
    element.removeChild(element.childNodes[index]);
  } else if (hasChanged(node, prevNode)) {
    element.replaceChild(
      createDocumentElement(node),
      element.childNodes[index]
    );
  } else if (isNode(node)) {
    updateAttributes(
      element.childNodes[index] as HTMLElement,
      node.props,
      prevNode.props
    );

    updateChildren(
      element,
      flattenChildren(node.children),
      flattenChildren(prevNode.children),
      index
    );
  } else if (element.childNodes[index].nodeValue !== node) {
    element.childNodes[index].nodeValue = node;
  }
};
