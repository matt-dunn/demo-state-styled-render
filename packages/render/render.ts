/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {Actions, Store} from "packages/state";

import {FC, Node} from "./types";
import {updateTree} from "./element";

type UseState = <S = any>(initialState: S) => [S, (newValue: S) => S];

type MxContainer = {
  index: number;
  state: any[];
  currentState: any;
  currentTree?: Node;
  renderTree: () => void;
  useState: UseState;
  render: <S>(store: Store<S>, actions: Actions) => (component: FC) => (mountPoint: HTMLElement) => void;
}

export const Mx: MxContainer = {
  index: 0,
  state: [],
  currentState: undefined,
  currentTree: undefined,
  renderTree: undefined as unknown as () => void,
  useState: <S = any>(initialState: S) => {
    const activeIndex = Mx.index;

    if (Mx.state[activeIndex] === undefined) {
      Mx.state[activeIndex] = initialState;
    }

    const currentState = Mx.state[activeIndex];

    const setter = (newValue: S) => {
      if (Mx.state[activeIndex] !== newValue) {
        Mx.state[activeIndex] = newValue;
        Mx.renderTree();
      }
      return newValue;
    };

    Mx.index++;

    return [currentState, setter];
  },
  render: (store, actions) => component => mountPoint => {
    const dispatchActions = Object.entries(actions).reduce((actions, [key, action]) => ({
      ...actions,
      [key]: (...args: any[]) => store.dispatch(action(...args))
    }), {});

    Mx.renderTree = () => {
      const currentTree = component({...Mx.currentState, ...dispatchActions});
      updateTree(mountPoint, currentTree, Mx.currentTree);

      Mx.currentTree = currentTree;
      Mx.index = 0;
    };

    Mx.currentState = store.getState();

    store.subscribe((state) => {
      Mx.currentState = state;
      Mx.renderTree();
    });

    return Mx.renderTree();
  }
};

export const useState = Mx.useState;
