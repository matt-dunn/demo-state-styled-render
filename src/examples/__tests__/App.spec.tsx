/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

/** @jsx jsx **/

import { jsx } from "packages/render";
import { mount } from "packages/render/test-utils";
import { createMockAction, MockAction } from "packages/state/test-utils/jest";

import { App } from "../App";
import { Actions as TodoActions } from "../todo";

describe("App: Example of using mount", () => {
  let createTodo: MockAction<TodoActions["createTodo"]>;
  let deleteTodo: MockAction<TodoActions["deleteTodo"]>;
  let updateTodo: MockAction<TodoActions["updateTodo"]>;

  beforeEach(() => {
    createTodo = createMockAction();
    deleteTodo = createMockAction();
    updateTodo = createMockAction();
  });

  it("should render component", async () => {
    const wrapper = mount(() => (
      <App
        todos={[]}
        createTodo={createTodo}
        deleteTodo={deleteTodo}
        updateTodo={updateTodo}
      />
    ));

    expect(wrapper).toMatchSnapshot();
  });
});
