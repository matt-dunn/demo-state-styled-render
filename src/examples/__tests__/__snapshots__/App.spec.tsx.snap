// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App: Example of using mount should render component 1`] = `
<div
  class="container-lg container-main"
>
  <header
    class=" ms__blmrZ"
  >
    <h1>
      Pure TypeScript implementation of styled JSX functional components and state management
    </h1>
    <aside
      class=" ms__lhcqzR"
    >
      <a
        href="https://matt-dunn.github.io/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <h2>
          Matt Dunn
        </h2>
        <p>
          Senior User Interface / Full Stack Developer
        </p>
      </a>
    </aside>
  </header>
  <article>
    <header>
      <p>
        Simple vanilla, dependency free implementation of:
      </p>
      <ul>
        <li>
          <a
            href="https://github.com/matt-dunn/demo-state-styled-render/tree/master/packages/render"
            rel="noopener noreferrer"
            target="_blank"
          >
            JSX functional component rendering with simple hooks
          </a>
        </li>
        <li>
          <a
            href="https://github.com/matt-dunn/demo-state-styled-render/tree/master/packages/state"
            rel="noopener noreferrer"
            target="_blank"
          >
            Redux style state management
          </a>
        </li>
        <li>
          <a
            href="https://github.com/matt-dunn/demo-state-styled-render/tree/master/packages/myStyled"
            rel="noopener noreferrer"
            target="_blank"
          >
            A Styled Component implementation
          </a>
        </li>
      </ul>
    </header>
    <p>
      Intended as an illustration of how the basics and internals of React / Redux / Styled Components work under the hood and
       
      <strong>
        not intended for production use
      </strong>
      .
    </p>
  </article>
  <main
    class="d-flex flex-md-row flex-column"
  >
    <section
      class="mr-md-2 mb-3 flex-grow-1 align-self-start col-md ms__brMznf"
    >
      <header
        class=" ms__jVxsUK"
      >
        <form
          class=" ms__iMxdBJ"
        >
          <label>
            Todo
            <input
              class="form-control"
              placeholder="Add new todo"
              type="text"
            />
          </label>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Add
          </button>
        </form>
        <p>
          You have 
          <strong>
            0
          </strong>
           todo
          s
        </p>
      </header>
      <main>
        <ul
          class=" ms__hUhVje"
        />
      </main>
    </section>
    <pre
      class="ml-md-2 flex-grow-1 col-md"
    >
      []
    </pre>
  </main>
</div>
`;
